stdlib/public/core/Codable.swift

```diff
@@ -3724,6 +3724,60 @@ public enum DecodingError: Error {
   }
 }
 
+extension DecodingError: CustomStringConvertible {
+  public var description: String {
+    switch self {
+    case .typeMismatch(let expectedType, let context):
+      return "Type mismatch: expected \(expectedType) at \(context.codingPath.readable)"
+    case .valueNotFound(let expectedType, let context):
+      return "Expected to find value of type \(expectedType), but value was missing at \(context.codingPath.readable)"
+    case .keyNotFound(let codingKey, let context):
+      return "Key not found: expected '\(codingKey.compactDescription)' at \(context.codingPath.readable)"
+    case .dataCorrupted(let context):
+      // The codingPath is empty here, so we defer to the underlying error.
+      var output = "Data corrupted: \(context.debugDescription)"
+      if let underlyingError = context.underlyingError {
+        if !output.hasSuffix(".") {
+          output.append(".")
+        }
+        output.append(" Underlying error: \(underlyingError)")
+      }
+      return output
+    }
+  }
+}
+
+private extension [any CodingKey] {
+  var readable: String {
+    self
+      .map { $0.compactDescription }
+      .joined(separator: "/")
+  }
+}
+
+private extension CodingKey {
+  /// Represents a coding key as either its integer representation, if available
+  /// (in square brackets), or its string representation. If there is no integer
+  /// value, and the string value happens to represent an integer, wrap it in
+  /// double quotes to avoid ambiguity.
+  var compactDescription: String {
+    if let intValue {
+      "[" + String(intValue) + "]"
+    } else {
+      // If the string key looks like an int, put quotes around it so it's clear
+      // that it's actually a string. If it's not a numeric string, don't bother
+      // because it's not ambiguous.
+      if let intyString = Int(stringValue) {
+        """
+        "\(intyString)"
+        """
+      } else {
+        stringValue
+      }
+    }
```

Comment:
How are we feeling about this style?

```suggestion
    intValue.map { "[" + String($0) + "]" } // wrap in square brackets
      ?? Int(stringValue).map { "\"\($0)\"" } // wrap in quotes
      ?? stringValue // fall back to plain string value
```